### PR TITLE
Test App Update for Ad ID Edge Identity Features

### PR DESCRIPTION
**Pull request by timkimadobe**
## Description

Update for the test app to be able to: 
- Fetch the ad ID from device (GAID) using the [Google Mobile Ads Lite SDK ](https://developers.google.com/admob/android/lite-sdk)
- Update advertising identifier using the Edge Identity extension (through MobileCore API)

The new functionality has been added under the Set Custom Identity view. The button language for fetching the GAID is not the same as the iOS test app, because the flow for ad ID permissions in Android is slightly different; Android automatically opts-in all users to ad ID tracking instead of a default opt-out status and asking permissions with a prompt like in iOS. 

Updates to build.gradle (:app):
- Kotlin version 1.3.72 -> 1.5.0 to enable using lifecycle-viewmodel-ktx
- play-services-ads-lite:20.6.0 to use Android ad ID provider APIs
- kotlinx-coroutines-android:1.6.0 to use Kotlin coroutines to handle async call to get ad ID
- lifecycle-viewmodel-ktx:2.5.0-alpha06 to handle managing coroutine scope lifecycle within ViewModel
- multidex:2.0.1 to fix exceeding 64k reference limit issue

## Related Issue

Please see Jira task: [AMSDK-11332 - (Android) Update Edge Identity test app for adid consent testing](https://jira.corp.adobe.com/browse/AMSDK-11332)

## Motivation and Context

With the new advertising identifier capabilities added to Edge Identity, the test app requires updates to be able to effectively test the new features using the Android Emulator.

## How Has This Been Tested?

Tested this implementation using the Android Emulator Pixel_3a_API_32_arm64-v8a, from a fresh install with a valid `LAUNCH_ENVIRONMENT_ID`. Navigated using hamburger menu to Set Custom Identity view, using the various new ad ID related buttons to test that the correct advertising identifier and consent events were being dispatched. Also verified the same in the View Identities view. Tested with the device level ad ID settings toggled on and off to verify correct all-zeros ad ID fetched when ad ID tracking disabled, and vice versa.

## Screenshots (if appropriate):
Note the new buttons for ad ID related functionality:
- Update Ad ID with Current GAID
Note: all Set Ad ID buttons use the `MobileCore.setAdvertisingIdentifier()` API, with special cases handled by dedicated button for ease of testing.
- Set Ad ID + textfield
- Set Ad ID as null
- Set Ad ID as all-zeros 
- Get Current Consents - uses `Consent.getConsents` API and prints log of result to console
Updated Set Custom Identity view:
Screen Shot 2022-04-26 at 1 56 32 PM

Screen Shot 2022-04-26 at 2 01 33 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (see related documentation PR https://github.com/adobe/aepsdk-edgeidentity-android/pull/58)
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by cacheung — COMMENTED** [review]
I tested the test app with feature-ad-id-test-app-update branch.    
I encounter the issue which I have to update Android studio to 2021.1.1. If using an older version of Android studio, getting an error when building: This version of the Android Support plugin for IntelliJ IDEA (or Android Studio) cannot open this project, please retry with version 2021.1.1 or newer.  Is it because the component we used in the Sample app causing that? Can we avoid asking for upgrading Android studio?

Also I am not able to get ECID after install the app to the device with Android studio 2011.1.1. 
I just installed the app, then click Get ECID.  Then get ECID not found error.

**Review by cacheung — COMMENTED** [review]
The test app runs good on M1 machine.  (we will need to upgrade version to 7.x).
For non-M1 machine, it can still use gradle version  6.1.1),  however, I am  not  able to  send Konductor call.  Getting  
2022-04-14 18:51:34.277 1684-1743/com.adobe.marketing.edge.identity.app D/AdobeExperienceSDK: Edge - EdgeExtension - Configuration is pending, couldn't process event at this time, waiting...

Need to validate more on a non-M1 machine.

**Review by cacheung — COMMENTED** [review]
The test app can be built and run on non-M1 mac as well now.

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt [hunk 1]
```diff
@@ -93,6 +104,44 @@ class CustomIdentityFragment : Fragment() {
...
 
+        // Button for Set Ad ID behavior
+        // Sets the advertising identifier set in the corresponding textfield using the MobileCore API
+        root.findViewById<Button>(R.id.btn_set_ad_id).setOnClickListener {
+            val adId: String? = sharedViewModel.adId.value
+            MobileCore.setAdvertisingIdentifier(adId)
+        }
+
+        // For details on implementation tips and differences between Google Play Services Ads vs AndroidX Ads,
+        // please see the project Documentation -> AEPEdgeIdentity.md : Test App -> GAID Implementation
```
Comment (cacheung) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt:115]: please see the project Documentation -> AEPEdgeIdentity.md : Android Test App -> GAID Implementations

**Review by cacheung — COMMENTED** [review]
It looks good.
Just do we need to make a note somewhere for M1, we need to install Andriod studio 2021  and upgrade gradle to 7.x version?

File: code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityExtensionTests.java [hunk 2]
```diff
@@ -629,39 +629,41 @@ public void test_processCachedEvents_processesWhenBooted() {
 	// ========================================================================================
 	// handleRequestContent
 	// ========================================================================================
-	// use callback to see if sharedstate is set properly
-	// similar to other ones where update identities
-	// if calling updateAdId directly, already passing in a callback
+
 	@Test
 	public void test_handleRequestContent_processesWhenBooted() {
```
Comment (emdobrin) [inline code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityExtensionTests.java:634]: Is this change intended or is it a result of the rebase? It seems off when comparing with the feature branch - https://github.com/adobe/aepsdk-edgeidentity-android/blob/feature/adIdConsent/code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityExtensionTests.java#L634

File: code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityStateTests.java [hunk 3]
```diff
@@ -542,4 +562,296 @@ public void testUpdateLegacyExperienceCloudId_notSetWhenExistingIsNull() {
...
+		state.getIdentityProperties().setECID(new ECID());
+
+		state.updateLegacyExperienceCloudId(null);
+
+		assertNull(state.getIdentityProperties().getECIDSecondary());
+		verify(mockSharedPreferenceEditor, Mockito.times(0)).apply();
+	}
+
+	@Test
+	public void testUpdateAdvertisingIdentifier_whenNull_thenChangedToValid() throws Exception {
```
Comment (emdobrin) [inline code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityStateTests.java:583]: Same comment on this file and EventUtilsTests and others, it seems like these changes were already merged, can you double check? You may need to merge the feature branch in this current branch for latest changes and merge any conflicts

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt [hunk 4]
```diff
@@ -99,6 +119,29 @@ class SharedViewModel : ViewModel() {
         _authenticatedStateId.value = value
     }
 
+    fun getGAID(applicationContext: Context) {
+        viewModelScope.launch(Dispatchers.Default) {
+            Log.d("Shared_View_Model", "Thread working in ${Thread.currentThread().name}")
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt:124]: nit: this seems like a debug log, let's remove

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt [hunk 5]
```diff
@@ -99,6 +119,29 @@ class SharedViewModel : ViewModel() {
...
+            try {
+                val idInfo = AdvertisingIdClient.getAdvertisingIdInfo(applicationContext)
+                if (idInfo.isLimitAdTrackingEnabled) {
+                    Log.d("Shared_View_Model", "Limit Ad Tracking is enabled by the user, setting ad ID to \"\"")
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt:128]: nit: For "SharedViewModel" you can use a single LOG_TAG constant for this class and reuse it everywhere you need it, this will help with renaming in the future.

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt [hunk 6]
```diff
@@ -93,6 +104,44 @@ class CustomIdentityFragment : Fragment() {
...
+        }
+
+        root.findViewById<Button>(R.id.btn_set_ad_id_all_zeros).setOnClickListener {
+            Log.d("Custom_Identity_Fragment", "Setting advertising identifier to: 00000000-0000-0000-0000-000000000000")
+            MobileCore.setAdvertisingIdentifier("00000000-0000-0000-0000-000000000000")
+        }
+
+        root.findViewById<Button>(R.id.btn_get_consents).setOnClickListener {
+            Consent.getConsents { consents ->
+                Log.d("Custom_Identity_Fragment", "Got Consents: $consents")
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt:141]: same comment here about the log tag, let's use a constant.

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt [hunk 7]
```diff
@@ -93,6 +104,44 @@ class CustomIdentityFragment : Fragment() {
...
+            val context = context
+            Log.d("Custom_Identity_Fragment", "context: $context, appContext: ${context?.applicationContext}")
+            if (context != null) {
+                // Implementation for google play services ads
+                // Logic implemented in ViewModel to leverage managed Kotlin coroutine scope from lifecycle-viewmodel-ktx
+                // lifecycle-viewmodel-ktx requires compiled sdk version 31+
+                GlobalScope.launch {
+                    sharedViewModel.getGAID(context.applicationContext)
+                }
+            }
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt:126]: Is this button intended to fetch the device advertising id or the adid currently set with the SDK?

**Review by emdobrin — COMMENTED** [review]
Seems like this branch is not in sync with feature/adIdConsent, can you please merge those changes in? Also, can you include a screenshot with the application UI changes in the PR description.

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt (hunk 5 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt]: Thank you! That sounds good. I also had a couple questions regarding `LOG_TAG`:
- Is there a preferred scope for the `LOG_TAG` declaration in this case (I put it in the `companion object`)? 
- Do you know if there is a naming convention for the `LOG_TAG` string? That is, camel case, underscores, etc.
- Is it ok that the LOG_TAG is scoped to the view model and not the test app level? Should the specific view model/fragment be part of the log string instead?

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt (hunk 6 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt]: Updated to use `LOG_TAG`

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityState.java [hunk 8]
```diff
@@ -163,18 +163,72 @@ void removeCustomerIdentifiers(final IdentityMap map) {
...
-	 *     <li>Update persistent storage with the new ad ID</li>
-	 *	   <li>Share the XDM state</li>
-	 * 	   <li>Dispatch consent event - only when ad ID changes from invalid/valid and vice versa</li>
-	 * </ul>
+	 * Construct the advertising identifier consent request event data using the provided consent value
+	 * @param consentVal the consent value defined by {@link IdentityConstants.XDMKeys.Consent#YES}
+	 *                   or {@link IdentityConstants.XDMKeys.Consent#NO}
+	 * @return the event data for advertising identifier consent request
+	 */
+	private Map<String, Object> buildConsentAdIdRequestData(final String consentVal) {
```
[the remark this answers is not in the document]
Comment (timkimadobe) [reply code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityState.java]: Sorry, I thought I merged the feature/adIdConsent branch into this one, but didn't realize my local version was pointing to my out of date fork branch of adIdConent; I've pulled the upstream into the fork and merged the latest changes into this testing branch! The updates from adIdConsent should be properly applied now

File: code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityExtensionTests.java (hunk 2 above)
Comment (timkimadobe) [reply code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityExtensionTests.java]: Should be fixed with latest adIdConsent merge! Thanks for pointing this out

**Review by timkimadobe — COMMENTED** [review]
Thanks for the review @emdobrin! The latest updates from feature/adIdConsent should be properly merged into this branch now! Just had a few small questions about LOG_TAG implementation, all other feedback should be addressed.

**Comment by timkimadobe** [comment]
> It looks good. Just do we need to make a note somewhere for M1, we need to install Andriod studio 2021 and upgrade gradle to 7.x version?

@cacheung thank you! I updated the README with notes for M1 development build tips addressing this issue in the documentation PR: https://github.com/adobe/aepsdk-edgeidentity-android/pull/58/commits/d3a476357da53235bdcf261144ff27431ff2f641

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt [hunk 9]
```diff
@@ -37,10 +42,17 @@ class CustomIdentityFragment : Fragment() {
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
+        val LOG_TAG = "Custom_Identity_Fragment"
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt:45]: This is a variable not a constant, you can use the same approach as per my other comment about constants.

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt [hunk 10]
```diff
@@ -93,6 +105,44 @@ class CustomIdentityFragment : Fragment() {
...
+        }
+
+        root.findViewById<Button>(R.id.btn_set_ad_id_null).setOnClickListener {
+            Log.d(LOG_TAG, "Setting advertising identifier to: null")
+            MobileCore.setAdvertisingIdentifier(null)
+        }
+
+        root.findViewById<Button>(R.id.btn_set_ad_id_all_zeros).setOnClickListener {
+            Log.d(LOG_TAG, "Setting advertising identifier to: 00000000-0000-0000-0000-000000000000")
+            MobileCore.setAdvertisingIdentifier("00000000-0000-0000-0000-000000000000")
```
Comment (emdobrin) [inline code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt:137]: nit: can use a constant for ALL_ZEROES_ADID and reuse it in logs too

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt (hunk 5 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/model/SharedViewModel.kt]: Thank you for all the information! This is super helpful; I've applied these changes to the logging system in the various views

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt (hunk 7 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt]: Implementation has been updated so that the getter function is a pure getter, and the returned ad ID is sent through the MobileCore.setAdvertisingIdentifier

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt (hunk 10 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt]: Usage updated!

File: code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt (hunk 9 above)
Comment (timkimadobe) [reply code/app/src/main/java/com/adobe/marketing/edge/identity/app/ui/CustomIdentityFragment.kt]: Usage updated!

**Review by timkimadobe — COMMENTED** [review]
Updated ad ID getter -> MobileCore.setAdvertisingIdentifier flow and log constants usage

**Review by emdobrin — APPROVED** [review]

File: code/app/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -15,13 +15,18 @@
...
+            android:exported="true">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
 
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <meta-data
+            android:name="com.google.android.gms.ads.APPLICATION_ID"
+            android:value="@string/gms_ads_app_id"/>
```
Comment (cacheung) [inline code/app/src/main/AndroidManifest.xml:28]: Without a valid  application id,  getting RuntimeException error
2022-05-03 14:28:32.407 32417-32417/com.adobe.marketing.edge.identity.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.adobe.marketing.edge.identity.app, PID: 32417
    java.lang.RuntimeException: Unable to get provider com.google.android.gms.ads.MobileAdsInitProvider: java.lang.IllegalStateException: 
    
    ******************************************************************************
    * Invalid application ID. Follow instructions here:                          *
    * https://googlemobileadssdk.page.link/admob-android-update-manifest         *
    * to find your app ID.                                                       *
    ******************************************************************************
    
    
        at android.app.ActivityThread.installProvider(ActivityThread.java:7244)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6780)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6697)
        at android.app.ActivityThread.access$1300(ActivityThread.java:237)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.IllegalStateException: 


Although it is expected. But getting  a  fatal exception error doesn't  give a good experience for the user.  How can we redirect the user to get setup their Application ID before running the app.

Comment (cacheung) [reply code/app/src/main/AndroidManifest.xml:28]: @timkimadobe Discussed this with Emilia, she suggested to add a step for  setup secret.xml with admob application id in the sample app [readme](https://github.com/adobe/aepsdk-edgeidentity-android#development)  -> Run demo application.
Once users go through the readme, then they should already have the id setup  and won't see the exception error.

Comment (emdobrin) [reply code/app/src/main/AndroidManifest.xml:28]: Will take this update in a separate PR.
